File: tray/src/theme.ts
import { applyDark } from '@taskscape/common-ui/theme';
import { getCurrentWindow } from '@tauri-apps/api/window';
import { api } from './api';

// The tray has no theme UI of its own; it mirrors the main app's Appearance
// preference. The effective light/dark is resolved in Rust (the shared setting,
// else the system appearance) and applied by toggling the `.dark` class the
// design tokens key their `dark:` variant off (see common-ui/styles/tokens.css).

/** Re-resolve the effective theme and apply it. Called on startup, on every
 *  reveal, and when the OS appearance flips. */
export async function refreshTheme(): Promise<void> {
  applyDark(await api.getDark().catch(() => false));
}

export function initTheme(): void {
  void refreshTheme();
  // Re-resolve when the OS appearance flips (covers the 'system' preference).
  void getCurrentWindow().onThemeChanged(() => void refreshTheme());
}

File: tray/src/theme.rs
//! Theme resolution. The tray has no theme UI of its own; it mirrors the main
//! app's Appearance preference (the shared "theme" setting), falling back to the
//! system appearance — resolved exactly like the old `get_dark` command, then
//! pushed into the Slint `Palette` global.

use slint::ComponentHandle;
use taskscape_common::Store;

use crate::{ctx, macos, Palette};

/// Resolve the effective dark/light: the shared setting if set, else the OS.
pub async fn resolve_dark(store: &Store) -> bool {
    match store.get_setting("theme").await.ok().flatten().as_deref() {
        Some("dark") => true,
        Some("light") => false,
        _ => macos::system_is_dark(),
    }
}

/// Re-resolve the theme off-thread and apply it to the bar (called on startup and
/// on every reveal).
pub fn refresh() {
    ctx::runtime().spawn(async {
        let dark = resolve_dark(&ctx::store()).await;
        ctx::with_ui(move |ui| ui.global::<Palette>().set_dark(dark));
    });
}
